src/lib/util/simh_tape_file.cpp

```diff
@@ -0,0 +1,624 @@
+// copyright-holders:Mietek Bak
+
+// best read together with SIMH magtape spec (rev 17 Jan 2022)
+// http://simh.trailing-edge.com/docs/simh_magtape.pdf
+
+#include <cassert>
+#include <cstring>
+#include <stdexcept>
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include "multibyte.h"
+#include "simh_tape_file.h"
+
+#if 0
+#define LOG(...)    fprintf(stderr, __VA_ARGS__)
+#else
+#define LOG(...)    ([](auto &&... args){})(__VA_ARGS__)
+#endif
+
+#define SIMH_TAPE_MARK                      0x00000000
+#define SIMH_ERASE_GAP                      0xfffffffe
+#define SIMH_HALF_GAP_FORWARD               0xfffeffff
+#define SIMH_END_OF_MEDIUM                  0xffffffff
+
+#define SIMH_GOOD_DATA_RECORD               0x0
+#define SIMH_PRIVATE_DATA_RECORD_1          0x1
+#define SIMH_PRIVATE_DATA_RECORD_2          0x2
+#define SIMH_PRIVATE_DATA_RECORD_3          0x3
+#define SIMH_PRIVATE_DATA_RECORD_4          0x4
+#define SIMH_PRIVATE_DATA_RECORD_5          0x5
+#define SIMH_PRIVATE_DATA_RECORD_6          0x6
+#define SIMH_PRIVATE_MARKER                 0x7
+#define SIMH_BAD_DATA_RECORD                0x8
+#define SIMH_RESERVED_DATA_RECORD_9         0x9
+#define SIMH_RESERVED_DATA_RECORD_A         0xa
+#define SIMH_RESERVED_DATA_RECORD_B         0xb
+#define SIMH_RESERVED_DATA_RECORD_C         0xc
+#define SIMH_RESERVED_DATA_RECORD_D         0xd
+#define SIMH_TAPE_DESCRIPTION_DATA_RECORD   0xe
+#define SIMH_RESERVED_MARKER                0xf
+
+inline const int get_marker_class(const int marker) { return marker >> 28; }
+inline const int get_marker_value(const int marker) { return marker & 0x0fffffff; }
+inline const bool is_marker_half_gap_reverse(const int marker) { return marker >= 0xffff0000 && marker <= 0xfffffffd; }
+
+//////////////////////////////////////////////////////////////////////////////
+
+// construction
+
+simh_tape_file::simh_tape_file(const std::string_view path, const bool create)
+	: m_page_size(-1)
+	, m_fd(-1)
+	, m_buf(nullptr)
+	, m_buf_len(0)
+	, m_read_only(false)
+	, m_pos(0)
+{
+	m_page_size = sysconf(_SC_PAGESIZE);
+	if (m_page_size == -1) // error: we failed to obtain page size
+		throw std::runtime_error(std::string("sysconf: ") + strerror(errno));
+
+	m_fd = open(std::string(path).c_str(), O_RDWR);
+	if (m_fd < 0) {
+		if (!create) { // retry as read-only if we're not creating new file
+			m_read_only = true;
+			m_fd = open(std::string(path).c_str(), O_RDONLY);
+		}
+		if (m_fd < 0) // error: we failed to open file
+			throw std::runtime_error(std::string("open: ") + strerror(errno));
+	}
+	if (!create) { // load existing file
+		const int result = lseek(m_fd, 0, SEEK_END);
+		if (result == -1) // error: we failed to seek to end of file
+			throw std::runtime_error(std::string("lseek: ") + strerror(errno));
+
+		m_buf_len = result;
+	}
+	else { // create new file
+		m_buf_len = 62914560; // we default to 60MB; TODO: allow specifying tape image size, once MAME supports it
```

Comment:
The specific mechanism isn't directly applicable here (@galibert may have some ideas), but MAME does support multiple options for creating floppy images (400K, 800K, and 1.44M for Apple 3.5" drives, for instance).  So you could choose Create and then have a choice of 60, 120, and 240 MB tapes, or whatever the standard sizes are there.

Reply:
It’s not clear to me how to provide such a choice within the confines of the MAME UI.  Please advise.

Reply:
Since there have been no further comments, I’ll leave this to future work.